content/300-guides/300-upgrade-guides/200-upgrading-versions/600-upgrading-to-prisma-5.mdx

```diff
@@ -0,0 +1,677 @@
+const prisma = new PrismaClient({
+  rejectOnNotFound: {
+    findUnique: true,
+  },
+})
+```
+
+You will need to update your codebase to use `findUniqueOrThrow` and `findFirstOrThrow` instead of `findUnique` and `findFirst`, depending on which calls you would like to throw.
+
+If instead you use a custom error handler with the `rejectOnNotFound` property, like these examples:
+
+```js
+// Example 3
+const prisma = new PrismaClient({
+  rejectOnNotFound: (err) => new Error('something'),
+})
+
+// Example 4
+const prisma = new PrismaClient({
+  rejectOnNotFound: {
+    findUnique: (err) => new Error('something'),
+  },
+})
+
+// Example 5
+const prisma = new PrismaClient({
+  rejectOnNotFound: {
+    findFirst: {
+      User: (err) => new Error('User error'),
+      Post: (err) => new Error('Post error'),
+    },
+    findUnique: {
+      User: (err) => new Error('User error'),
+      Post: (err) => new Error('Post error'),
+    },
+  },
+})
+```
+
+You will need to update your method usage to `...OrThrow` and then use a [client extension](/concepts/components/prisma-client/client-extensions) in order to get the same behavior. As an example, the following extension would give the same behavior in Prisma 5 that `Example 5` gave in Prisma 4 and lower.
+
+```js
+import { PrismaClient } from '@prisma/client';
+
+const customErrorFunc = async (model, query, args) => {
+  try {
+    await query(args)
+  } catch (error: any) {
+    if (error.code === 'P2025') {
+      throw new Error(`${model} error`)
+    }
+    throw error;
+  }
+}
+
+const prisma = (new PrismaClient()).$extends({
+  query: {
+    user: {
+      async findFirstOrThrow({ model, query, args }) {
+        return await customErrorFunc(model, query, args)
+      },
+      async findUniqueOrThrow({ model, query, args }) {
+        return await customErrorFunc(model, query, args)
+      },
+    },
+    post: {
+      async findFirstOrThrow({ model, query, args }) {
+        return await customErrorFunc(model, query, args)
+      },
+      async findUniqueOrThrow({ model, query, args }) {
+        return await customErrorFunc(model, query, args)
+      },
+    },
+  },
+})
+```
+
+### <inlinecode>jsonProtocol</inlinecode> out of Preview!
+
+The `jsonProtocol` preview feature is now Generally Available. When upgrading to Prisma 5, make sure to remove `jsonProtocol` from your preview features, if added.
+
+Prisma 4 and lower:
+
+```prisma
+generator client {
+  provider        = "prisma-client-js"
+  previewFeatures = ["jsonProtocol"]
+}
+```
+
+Prisma 5:
+
+```prisma
+generator client {
+  provider = "prisma-client-js"
+}
+```
+
+#### Improved error messages
+
+Due to the switch to the new protocol, several error messages have been improved. For example, the following error message in Prisma 4 and below
+
+```terminal
+Argument id for where.id must not be null. Please use undefined instead.
+```
+
+becomes the following in Prisma 5:
+
+```terminal
+Argument `id` must not be null.
+```
```

Comment:
I don't see how this example is an improvement — it's just different, but it's a neutral change IMO. We could find examples where they have been improved with the new protocol, but there are also cases where they've become worse for now, and we'll look into improving them later.

So I don't think it's safe or fair to say we have "improved" the error messages — they just became different because of the new protocol and validation layer, and that's all.

Reply:
(Joël linked to an example of a real improvement in a comment above: https://prisma-company.slack.com/archives/C03MRGP7WLB/p1689065182410019?thread_ts=1689036527.158899&cid=C03MRGP7WLB)

Reply:
Added a proper improvement. Thank you!